### PR TITLE
bound string buffer byteswap by buffer_size in ByteSwapBuffer

### PR DESCRIPTION
**Pull request by madib06ops**
The TensorType_STRING case in ByteSwapBuffer reads the string count from the first int32 of the buffer and swaps num_of_strings + 2 words, ignoring buffer_size, so a crafted TFLite model whose string tensor declares a count larger than its buffer overruns that buffer when loaded on a big-endian host. It is reachable from untrusted model data via the non-verifying BuildFromBuffer/BuildFromFile paths, which run ByteConvertModel without tflite::VerifyModelBuffer. Cap the number of int32 words to buffer_size / sizeof(int32_t), matching the other dtype cases, and skip buffers too small to hold the count.

### CODE REVIEW

File: tensorflow/compiler/mlir/lite/core/model_builder_base.h [hunk 1]
```diff
@@ -220,10 +220,20 @@ class FlatBufferModelBase {
...
+        const size_t max_words = buffer_size / sizeof(int32_t);
+        if (max_words == 0) break;
         auto bp = reinterpret_cast<int32_t*>(buffer);
         int num_of_strings =
             from_big_endian ? bp[0] : flatbuffers::EndianSwap(bp[0]);
-        for (int i = 0; i < num_of_strings + 2; i++)
+        size_t words =
+            num_of_strings < 0
+                ? max_words
+                : std::min(static_cast<size_t>(num_of_strings) + 2, max_words);
```
[the remark this answers is not in the document]
Comment (madib06ops) [reply tensorflow/compiler/mlir/lite/core/model_builder_base.h:239]: Good catch. Swapping the whole buffer on a negative count was wrong on both counts, so I changed the negative branch to swap only the count word. Also added a ByteSwapStringBufferHandlesNegativeCount test alongside the existing one to lock that path in.

**Review by dmiltr3 — CHANGES_REQUESTED** [review]
Thank you for this contribution! The changes look solid and address the buffer overflow vulnerability in `ByteSwapBuffer` safely.

Here are a few observations and suggestions:

### 1. Robustness of the Fix
The introduction of `max_words` and the early break if `max_words == 0` correctly prevents reading or writing past the end of buffers too small to hold even the string count.

### 2. Handling Negative Counts
The handling of negative counts by capping `words` to 1 is a clever fix to avoid interpreting negative lengths as large unsigned bounds (which would otherwise happen if naively casting to `size_t` or using standard mins without bounds).

### 3. Additional Test Case (Suggested)
The tests you added cover Oversized and Negative counts well. It might be beneficial to also include a test case for undersized buffers (e.g., smaller than `sizeof(int32_t)`) to verify that the `max_words == 0` condition is handled gracefully without reading `bp[0]`.

Here is a test case you could add to `tensorflow/lite/core/model_test.cc` (inside the `#if FLATBUFFERS_LITTLEENDIAN == 0` block, as these methods are only used/compiled on Big Endian platforms):

```cpp
#if FLATBUFFERS_LITTLEENDIAN == 0
TEST(BasicFlatBufferModel, ByteSwapStringBufferTooSmall) {
  // 2 bytes buffer, smaller than sizeof(int32_t)
  std::vector<uint8_t> storage = {0x01, 0x02}; 
  const size_t buffer_size = storage.size();
  
  // Should break early and change nothing
  FlatBufferModel::ByteSwapBuffer(
      static_cast<int8_t>(TensorType_STRING), buffer_size,
      storage.data(), /*from_big_endian=*/true);
      
  EXPECT_EQ(storage[0], 0x01);
  EXPECT_EQ(storage[1], 0x02);
}
#endif
```

### 4. Safety Improvement (Optional)
While not strictly blocking (as existing code doesn't check it), it is generally good practice to add a null pointer check for `buffer` at the beginning of `ByteSwapBuffer` to prevent potential crashes if the method is called with invalid pointers in other contexts.

```cpp
if (buffer == nullptr) return;
```

**Comment by madib06ops** [comment]
Both good calls. Added the ByteSwapStringBufferTooSmall test for the sub-int32 case so the max_words == 0 early break stays covered, and put a nullptr guard at the top of ByteSwapBuffer. Pushed in 91f05329c.

**Review by dmiltr3 — APPROVED** [review]
Looks good to me! All previous review comments have been addressed, edge cases (undersized buffer, negative count, buffer boundary) are well covered with unit tests, and all CI checks pass.
